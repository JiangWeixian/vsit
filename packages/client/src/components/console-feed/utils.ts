import { css } from '@codemirror/lang-css'
import { html } from '@codemirror/lang-html'
import { javascript } from '@codemirror/lang-javascript'
import { HighlightStyle } from '@codemirror/language'
import { tags } from '@lezer/highlight'

import type { LanguageSupport } from '@codemirror/language'

// TODO: add types
type CustomLanguage = any

export const getLanguageFromFile = (
  filePath: string | undefined,
  fileType: string | undefined,
  additionalLanguages: CustomLanguage[],
): string => {
  if (!filePath && !fileType) {
    return 'javascript'
  }

  let extension = fileType
  if (!extension && filePath) {
    const extensionDotIndex = filePath.lastIndexOf('.')
    extension = filePath.slice(extensionDotIndex + 1)
  }

  for (const additionalLanguage of additionalLanguages) {
    if (
      extension === additionalLanguage.name
      || additionalLanguage.extensions.includes(extension || '')
    ) {
      return additionalLanguage.name
    }
  }

  switch (extension) {
    case 'ts':
    case 'tsx':
      return 'typescript'
    case 'html':
    case 'svelte':
    case 'vue':
    case 'astro':
      return 'html'
    case 'css':
    case 'less':
    case 'scss':
      return 'css'
    case 'js':
    case 'jsx':
    case 'json':
    default:
      return 'javascript'
  }
}

export const getCodeMirrorLanguage = (
  extension: string,
  additionalLanguages: CustomLanguage[],
): LanguageSupport => {
  const options: Record<string, LanguageSupport> = {
    javascript: javascript({ jsx: true, typescript: false }),
    typescript: javascript({ jsx: true, typescript: true }),
    html: html(),
    css: css(),
  }

  for (const additionalLanguage of additionalLanguages) {
    if (extension === additionalLanguage.name) {
      return additionalLanguage.language
    }
  }

  return options[extension as keyof typeof options]
}

// Based on: https://github.com/uiwjs/react-codemirror/blob/master/themes/atomone/src/index.ts
// https://github.com/codesandbox/sandpack/blob/main/sandpack-themes/src/atomDark.ts
export const getSyntaxHighlight = (): HighlightStyle =>
  HighlightStyle.define([
    {
      tag: [tags.function(tags.variableName), tags.function(tags.propertyName), tags.url, tags.processingInstruction],
      color: 'hsl(207, 82%, 66%)',
    },
    { tag: [tags.tagName, tags.heading], color: '#e06c75' },
    { tag: [tags.comment, tags.quote], color: '#757575' },
    { tag: tags.comment, fontStyle: 'italic' },
    { tag: [tags.propertyName], color: 'hsl(220, 14%, 71%)' },
    { tag: [tags.atom, tags.number, tags.bool, tags.attributeName], color: 'hsl( 29, 54%, 61%)' },
    { tag: tags.className, color: 'hsl( 39, 67%, 69%)' },
    { tag: tags.keyword, color: 'hsl(286, 60%, 67%)' },
    { tag: [tags.string, tags.regexp, tags.special(tags.propertyName)], color: '#98c379' },

    { tag: tags.link, textDecoration: 'underline' },
    { tag: tags.emphasis, fontStyle: 'italic' },
    { tag: tags.strong, fontWeight: 'bold' },

    {
      // Standard tags, e.g <h1 />
      tag: tags.standard(tags.tagName),
      color: '#e06c75',
    },
    {
      tag: [
        // Highlight function call
        tags.function(tags.variableName),

        // Highlight function definition differently (eg: functional component def in React)
        tags.definition(tags.function(tags.variableName)),

        // "Custom tags", meaning React component
        tags.tagName,
      ],
      color: '#62aeef',
    },
    {
      tag: [tags.literal, tags.inserted],
      color: '#98c379',
    },
    {
      tag: tags.punctuation,
      color: '#a8b1c2',
    },
  ])
