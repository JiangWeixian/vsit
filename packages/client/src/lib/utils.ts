
export const withQuery = (pathname: string) => {
  const params = { t: Date.now().toString() }
  const search = new URLSearchParams(params)
  return `${pathname}?${search}`
}