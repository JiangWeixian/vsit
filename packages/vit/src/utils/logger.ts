// refs: https://github.com/vuejs/vue-cli/blob/v2/lib/logger.js
import { format } from 'node:util'

import pc from 'picocolors'

/**
 * Prefix.
 */
const sep = pc.gray('Â·')

/**
 * Log a `message` to the console.
 *
 * @param {String} message
 */

const log = (...args: [any, ...any[]]) => {
  const msg = format.apply(format, args)
  console.log(pc.bgBlue(pc.black(' info ')), sep, msg)
}

/**
 * Log an error `message` to the console and exit.
 *
 * @param {String} message
 */

const fatal = (...args: [any, ...any[]]) => {
  if (args[0] instanceof Error) {
    args[0] = args[0].message.trim()
  }
  const msg = format.apply(format, args)
  console.error(pc.bgRed(pc.black(' failed ')), sep, msg)
  process.exit(1)
}

/**
 * Log a success `message` to the console and exit.
 *
 * @param {String} message
 */

const success = (...args: [any, ...any[]]) => {
  const msg = format.apply(format, args)
  console.log(pc.bgRed(pc.black(' success ')), sep, msg)
}

export default {
  log,
  fatal,
  success,
}
