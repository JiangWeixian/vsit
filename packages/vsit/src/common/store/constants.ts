import os from 'node:os'
import path from 'node:path'

const STORE_DIR = '.vsit-store'
export const STORE_PATH = path.join(os.homedir(), STORE_DIR)
export const LOCK_FILE = 'vist-lock.yaml'
export const STORE_PACKAGES_DIR = 'packages'
