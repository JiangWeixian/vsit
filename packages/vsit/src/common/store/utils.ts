import { createHash } from 'node:crypto'

export const computeCacheKey = (url: string) => {
  const hash = createHash('sha256').update(url).digest('hex')
  return hash
}
