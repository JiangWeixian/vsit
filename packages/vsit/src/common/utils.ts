import { join } from 'node:path'

import { withoutLeadingSlash, withoutTrailingSlash } from 'ufo'

import { pkgRoot } from './path'
import {
  ESM_HOST,
  ESMSH_HTTP_RE,
  ESMSH_HTTP_SUB_RE,
  ESMSH_PROTOCOL,
  ESMSH_PROTOCOL_RE,
  NULL_BYTE,
  NULL_BYTE_PLACEHOLDER,
  VALID_ID_PREFIX,
  VIRTUAL_RE,
} from './resolver/constants'
import { isEsmSh } from './resolver/is'
import { computeCacheKey } from './store/utils'

import type { ModuleNode } from 'vite'
import type { Package } from './store/persist-cache'

// '\0' tell vite to not resolve this id via internal node resolver algorithm
export const wrapId = (id: string) => {
  // esm.sh:xxx -> https://esm.sh/xxx
  if (ESMSH_PROTOCOL_RE.test(id)) {
    return `\0${id.replace(ESMSH_PROTOCOL_RE, ESM_HOST)}`
  }
  // /v126/xx -> https://esm.sh/v126/xx
  if (ESMSH_HTTP_SUB_RE.test(id)) {
    return `\0${ESM_HOST}${withoutLeadingSlash(id)}`
  }
  return id
}

export const unwrapId = (id: string) => {
  let stripId = id.replace(VIRTUAL_RE, '')
  // unwrap
  // https:/esm.sh -> https://esm.sh
  // esm.sh: -> https://esm.sh
  stripId = stripId
    .replace(VALID_ID_PREFIX, '')
    .replace(NULL_BYTE_PLACEHOLDER, '')
    .replace(NULL_BYTE, '')
    .replace(ESMSH_HTTP_RE, withoutTrailingSlash(ESM_HOST))
    .replace(ESMSH_PROTOCOL_RE, ESM_HOST)
  return stripId
}

// TODO: should use es-lexer and magic-string?
export const wrapCode = (code: string) => {
  return code.replace(new RegExp(ESM_HOST, 'gi'), ESMSH_PROTOCOL)
}

export const injectConsoleHook = (content: string) => {
  const entryOfVit = process.env.TEST ? 'vsit' : join(pkgRoot, 'dist/node.mjs')
  return `
import { consolehook } from "${entryOfVit}"
globalThis.__hook(consolehook, (log) => {
  console.log(log)
  globalThis.__viteDevServer.ws.send({
    type: 'custom',
    data: globalThis.__encode(log),
    event: 'vsit:custom',
  })
})
${content}
`
}

export const parseDeps = (deps: string[] = []) => {
  return deps
    .map((dep) => {
      return unwrapId(dep)
    })
    .filter((dep) => {
      return isEsmSh(dep)
    })
}

const createPackage = (url: string, deps: string[]): Record<string, Package> => {
  if (!url) {
    return {}
  }
  const id = computeCacheKey(url)
  return {
    [id]: {
      id,
      url,
      deps,
    },
  }
}

export const parseModulesDeps = (m?: ModuleNode): Record<string, Package> => {
  if (!m || !m.id) {
    return {}
  }
  const id = unwrapId(m.id)
  const deps = parseDeps(m.ssrTransformResult?.deps)
  let records = id && isEsmSh(id) && deps.length
    ? createPackage(id, deps)
    : {}
  m.importedModules.forEach((importedModule) => {
    const result = parseModulesDeps(importedModule)
    records = {
      ...records,
      ...result,
    }
  })
  return records
}
