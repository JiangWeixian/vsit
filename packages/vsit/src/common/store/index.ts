import { createHash } from 'node:crypto'

import { fetch } from 'ofetch'

import { PersistCache } from './persist-cache'
import { debug } from '@/common/log'

export const createStore = async () => {
  const pool = new Map<string, Promise<string>>()
  const cacheManager = await PersistCache.create()
  const createInstance = (id: string, url: string, options?: RequestInit) => {
    const promise = (async () => {
      try {
        debug.store('start fetch %s', url)
        return fetch(url, options)
          .then(async (res) => {
            const content = await res.text()
            cacheManager.saveCache(url, content)
            pool.delete(id)
            return content
          })
      } catch (e) {
        // TODO: add retry
        console.error(e)
        return Promise.reject(e)
      }
    })()
    return promise
  }
  return {
    cache: cacheManager,
    async clear(url: string) {
      const hash = createHash('sha256').update(url).digest('hex')
      pool.delete(hash)
    },
    async fetch(url: string, options?: RequestInit) {
      const hash = createHash('sha256').update(url).digest('hex')
      const cache = await cacheManager.getCache(url)
      if (cache) {
        debug.store('load cache %s', url)
        return Promise.resolve(cache)
      }
      let instance = pool.get(hash)
      if (instance) {
        return instance
      }
      instance = createInstance(hash, url, options)
      instance && pool.set(hash, instance)
      return instance
    },
  }
}
