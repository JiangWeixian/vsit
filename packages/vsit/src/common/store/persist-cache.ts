import { resolve } from 'node:path'

import {
  existsSync,
  outputFile,
  readFile,
} from 'fs-extra'
import readYaml from 'read-yaml-file'
import writeYaml from 'write-yaml-file'

import { version } from '../../../package.json'
import {
  LOCK_FILE,
  STORE_PACKAGES_DIR,
  STORE_PATH,
} from './constants'
import { computeCacheKey } from './utils'
import { debug } from '@/common/log'

interface Options {
  storePath?: string
}

interface ResolvedLockFileOptions {
  storePath: string
  lockFilePath: string
}

export interface Package {
  id: string
  url: string
  deps?: string[]
}

interface LockFileYaml {
  version: string
  packages?: Record<string, Package>
}

export class LockFile {
  options: ResolvedLockFileOptions
  lockFile: LockFileYaml = { version }
  constructor(options: Options) {
    this.options = this.resolveOptions(options)
    debug.store('resolved lock-file options %o', this.options)
  }

  resolveOptions(options: Options): ResolvedLockFileOptions {
    const resolvedStorePath = options.storePath ?? STORE_PATH
    return {
      storePath: resolvedStorePath,
      lockFilePath: resolve(resolvedStorePath, LOCK_FILE),
    }
  }

  async init() {
    if (!existsSync(this.options.lockFilePath)) {
      await writeYaml(this.options.lockFilePath, { version })
    }
    await this.read()
  }

  async read(): Promise<LockFileYaml> {
    this.lockFile = await readYaml(this.options.lockFilePath)
    return this.lockFile as LockFileYaml
  }

  async write(data: LockFileYaml): Promise<void> {
    await writeYaml(this.options.lockFilePath, data)
  }

  async save() {
    await writeYaml(this.options.lockFilePath, this.lockFile)
  }

  async writePackage(url: string, deps: string[] = []) {
    const id = computeCacheKey(url)
    this.lockFile.packages = {
      ...this.lockFile.packages,
      [id]: {
        id,
        url,
        deps,
      },
    }
    await this.save()
  }

  async writePackages(packages: Record<string, Package> = {}) {
    this.lockFile.packages = {
      ...this.lockFile.packages,
      ...packages,
    }
    await this.save()
  }

  static async create(options: Options) {
    const instance = new LockFile(options)
    await instance.init()
    return instance
  }
}

interface ResolvedCacheOptions {
  storePath: string
  packagesPath: string
}

export class PersistCache {
  options: ResolvedCacheOptions
  lockFile?: LockFile
  constructor(options: Options = { storePath: STORE_PATH }) {
    this.options = this.resolveOptions(options)
    this.lockFile = undefined
  }

  resolveOptions(options: Options): ResolvedCacheOptions {
    const resolvedStorePath = options.storePath ?? STORE_PATH
    return {
      storePath: resolvedStorePath,
      packagesPath: resolve(resolvedStorePath, STORE_PACKAGES_DIR),
    }
  }

  static async create(options: Options = { storePath: STORE_PATH }) {
    const instance = new PersistCache(options)
    instance.lockFile = await LockFile.create(options)
    return instance
  }

  async writePackages(packages: Record<string, Package> = {}) {
    this.lockFile?.writePackages(packages)
  }

  async getCache(url: string) {
    const id = computeCacheKey(url)
    const path = resolve(this.options.packagesPath, id)
    if (existsSync(path)) {
      const content = (await readFile(path)).toString('utf-8')
      return content
    }
    return ''
  }

  async saveCache(url: string, content: string) {
    const id = computeCacheKey(url)
    const path = resolve(this.options.packagesPath, id)
    await this.lockFile?.writePackage(url)
    await outputFile(path, content)
  }
}

export const createPersistCache = () => {}
