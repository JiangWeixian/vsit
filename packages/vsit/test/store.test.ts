import { existsSync, removeSync } from 'fs-extra'
import { temporaryDirectory } from 'tempy'
import {
  afterAll,
  beforeAll,
  describe,
  expect,
  it,
} from 'vitest'

import { LockFile } from '@/common/store/persist-cache'
import { computeCacheKey } from '@/common/store/utils'

describe('lock file', () => {
  let lockFile: LockFile
  const storePath = temporaryDirectory()
  beforeAll(async () => {
    lockFile = await LockFile.create({
      storePath,
    })
  })
  afterAll(() => {
    removeSync(storePath)
  })
  it('lock-file should exit after init', async () => {
    expect(existsSync(lockFile.options.lockFilePath)).toBe(true)
  })
  it('read lock-file should work', async () => {
    const content = await lockFile.read()
    expect(content).toBeDefined()
  })
  it('write lock-file should work', async () => {
    await lockFile.write({ version: '1.0.0' })
    const content = await lockFile.read()
    expect(content.version).toBe('1.0.0')
  })
  it('write packages should work', async () => {
    const url = 'https://esm.sh/lodash'
    await lockFile.writePackage('https://esm.sh/lodash')
    const key = computeCacheKey(url)
    const content = await lockFile.read()
    expect(content.packages?.[key]).toMatchSnapshot()
  })
  it('write packages should work', async () => {
    await lockFile.writePackages({
      'https://esm.sh/strip': {
        id: 'https://esm.sh/lodash',
        url: 'https://esm.sh/lodash',
        deps: [],
      },
    })
    const content = await lockFile.read()
    expect(content.packages?.['https://esm.sh/strip']).toMatchSnapshot()
  })
})
