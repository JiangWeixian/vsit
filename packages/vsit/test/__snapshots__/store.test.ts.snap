// Vitest Snapshot v1

exports[`lock file > write packages should work 1`] = `
{
  "deps": [],
  "id": "210a361251f5fc11991e0e11ff17b5b543b3d992ab4fb473782e3589571b16ad",
  "url": "https://esm.sh/lodash",
}
`;

exports[`lock file > write packages should work 2`] = `
{
  "deps": [],
  "id": "https://esm.sh/lodash",
  "url": "https://esm.sh/lodash",
}
`;
